Add reduce variants for `scatter` operation.

ysiraichi opened the pull request:
Fixes #56463 #56464

### Summary
- Add reduce variants for `scatter` in both _native_functions.yaml_ and _TensorAdvancedIndexing.cpp_
- Add `OpInfo` tests and reduce tests in _test_torch.py_
- Fix default reduce argument for `scatter_` in __tensor_docs.py_

ysiraichi committed:
Add `reduce` parameter for out-of-place `scatter` operation.

ysiraichi committed:
Add tests for out-of-place `scatter` with `reduce` argument.

ysiraichi committed:
Add `scatter` reduce variant for `OpInfo` tests.

ysiraichi committed:
Add functional variant for `scatter`.

ysiraichi committed:
Tests passing.

ysiraichi committed:
Fix `scatter_` doc.

ysiraichi committed:
Unshadow imported reduce.

ysiraichi commented:
I believe these changes have nothing to do with the CI failure.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4877,11 +4877,21 @@
+- func: scatter.reduce(Tensor self, int dim, Tensor index, Tensor src, *, str reduce) -> Tensor
+  variants: function, method
+  dispatch:
+    CPU, CUDA: scatter_reduce
```
Should there also be an out variant of this?

ysiraichi replied:
I don't see why not! I will add them.

ezyang replied:
If so, consider making it structured! https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4877,11 +4877,21 @@
+- func: scatter.value_reduce(Tensor self, int dim, Tensor index, Scalar value, *, str reduce) -> Tensor
+  variants: function, method
+  dispatch:
+    CPU, CUDA: scatter_scalar_reduce
```
I think you are going to need to add `derivatives.yaml` entries for this

ezyang commented:
This all looks very reasonable but I am not sure autograd is going to be working. cc @mruberry for testing related changes

ysiraichi committed:
Merge reduce args as optional and add `out` variants.

ysiraichi committed:
Add derivative for merged function.

ysiraichi committed:
Change sample input generation.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2982,7 +2980,22 @@ def _gather(shape, index_dim, max_indices):
+    for tensor, args in test_cases:
+        samples.append(SampleInput(tensor, args=args))
+
+        if not requires_grad:
```
Interesting! So these args are only supported when the input doesn't require grad?

ysiraichi replied:
Yes. In fact, there were not _derivatives.yaml_ items for the reduce variants. I think I can implement them (not really confident about my math skills, though). As far as I understand, it could be implemented like:

- **add:** (same as `scatter_add`)
```python
> self_grad = grad
> src_grad = grad.gather(dim, index)
```
- **multiply:**
```python
> self_grad = grad.scatter(dim, index, src, 'multiply')
> src_grad = grad.gather(dim, index).mul(self.gather(dim, index))
```

What do you think?

albanD replied:
Yes sure that could be a different PR for sure.
Note that the current update is kind of problematic. So some update needs to be done to be sure we don't regress existing functions.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5085,6 +5098,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            dtypes=all_types_and_complex_and(torch.bool, torch.half),
            dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
            sample_inputs_func=sample_inputs_scatter,
+           supports_complex_autograd=False,
```
This will cause a logical merge conflict because this metadata has been generalized here:

https://github.com/pytorch/pytorch/blob/151e81b7bc0ae430c95045c36928e1198b0131be/torch/testing/_internal/common_methods_invocations.py#L161

You want to list all the forward types except the complex types.

I'm confused, however, why does scatter no longer support complex autograd? Or did it never support it and this was improperly tested?

ysiraichi replied:
I think, for some reason, it never supported... At least, it is not inside the list [`GRADIENT_IMPLEMENTED_FOR_COMPLEX`](https://github.com/pytorch/pytorch/blob/9ec6883442df8ead765941441854a5518e35c69c/tools/autograd/gen_variable_type.py#L81). I guess, I can add it there. Not sure about whether it's correct or not, though.

mruberry replied:
cc @anjali411 to take a look

mruberry reviewed (changes requested):
Hey @ysiraichi -- just a couple questions on testing and a rebase is needed to avoid a logical merge conflict with complex autograd metadata

ysiraichi committed:
Port `scatter` to structured kernel.

ysiraichi committed:
Add support for complex gradient.

ysiraichi committed:
Remove unnecessarily added include.

ysiraichi committed:
Restore docs (now, we support an optional reduce arg).

ysiraichi committed:
Remove trailing spaces.

ezyang commented:
this is gonna need XLA side changes cc @JackCaoG

JackCaoG commented:
I will work on the fix and update here when it is ready

mruberry commented:
Some of the ONNX failures might be related to this PR:

```
May 05 16:33:33 =================================== FAILURES ===================================
May 05 16:33:33 _________________________ TestONNXRuntime.test_scatter _________________________
May 05 16:33:33 Traceback (most recent call last):
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_common.py", line 49, in wrapper
May 05 16:33:33     return func(self)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 3166, in test_scatter
May 05 16:33:33     self.run_test(ScatterModel(), input=(input, indices, values))
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 254, in run_test
May 05 16:33:33     _run_test(script_model)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 251, in _run_test
May 05 16:33:33     training=training)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 133, in run_model_test
May 05 16:33:33     onnx_shape_inference=self.onnx_shape_inference)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 61, in convert_to_onnx
May 05 16:33:33     onnx_shape_inference=onnx_shape_inference)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 28, in _export
May 05 16:33:33     result = utils._export(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 682, in _export
May 05 16:33:33     dynamic_axes=dynamic_axes)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 454, in _model_to_graph
May 05 16:33:33     module=module)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 200, in _optimize_graph
May 05 16:33:33     graph = torch._C._jit_pass_onnx(graph, operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
May 05 16:33:33     return utils._run_symbolic_function(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 981, in _run_symbolic_function
May 05 16:33:33     return symbolic_fn(g, *inputs, **attrs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/symbolic_helper.py", line 134, in wrapper
May 05 16:33:33     assert len(arg_descriptors) >= len(args)
May 05 16:33:33 AssertionError
May 05 16:33:33 ___________________ TestONNXRuntime.test_scatter_with_scalar ___________________
May 05 16:33:33 Traceback (most recent call last):
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_common.py", line 49, in wrapper
May 05 16:33:33     return func(self)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 3141, in test_scatter_with_scalar
May 05 16:33:33     self.run_test(ScatterModel(), input=(input, indices))
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 254, in run_test
May 05 16:33:33     _run_test(script_model)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 251, in _run_test
May 05 16:33:33     training=training)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 133, in run_model_test
May 05 16:33:33     onnx_shape_inference=self.onnx_shape_inference)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 61, in convert_to_onnx
May 05 16:33:33     onnx_shape_inference=onnx_shape_inference)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 28, in _export
May 05 16:33:33     result = utils._export(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 682, in _export
May 05 16:33:33     dynamic_axes=dynamic_axes)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 454, in _model_to_graph
May 05 16:33:33     module=module)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 200, in _optimize_graph
May 05 16:33:33     graph = torch._C._jit_pass_onnx(graph, operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
May 05 16:33:33     return utils._run_symbolic_function(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 981, in _run_symbolic_function
May 05 16:33:33     return symbolic_fn(g, *inputs, **attrs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/symbolic_helper.py", line 134, in wrapper
May 05 16:33:33     assert len(arg_descriptors) >= len(args)
May 05 16:33:33 AssertionError
May 05 16:33:33 ___________ TestONNXRuntime.test_scatter_with_scalar_different_types ___________
May 05 16:33:33 Traceback (most recent call last):
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_common.py", line 49, in wrapper
May 05 16:33:33     return func(self)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 3155, in test_scatter_with_scalar_different_types
May 05 16:33:33     self.run_test(ScatterModel(), input=(input, indices))
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 254, in run_test
May 05 16:33:33     _run_test(script_model)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 251, in _run_test
May 05 16:33:33     training=training)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 133, in run_model_test
May 05 16:33:33     onnx_shape_inference=self.onnx_shape_inference)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_onnxruntime.py", line 61, in convert_to_onnx
May 05 16:33:33     onnx_shape_inference=onnx_shape_inference)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 28, in _export
May 05 16:33:33     result = utils._export(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 682, in _export
May 05 16:33:33     dynamic_axes=dynamic_axes)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 454, in _model_to_graph
May 05 16:33:33     module=module)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 200, in _optimize_graph
May 05 16:33:33     graph = torch._C._jit_pass_onnx(graph, operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
May 05 16:33:33     return utils._run_symbolic_function(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 981, in _run_symbolic_function
May 05 16:33:33     return symbolic_fn(g, *inputs, **attrs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/symbolic_helper.py", line 134, in wrapper
May 05 16:33:33     assert len(arg_descriptors) >= len(args)
May 05 16:33:33 AssertionError
May 05 16:33:33 __________________ TestCaffe2BackendEmbed_opset9.test_scatter __________________
May 05 16:33:33 Traceback (most recent call last):
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_common.py", line 49, in wrapper
May 05 16:33:33     return func(self)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 1616, in test_scatter
May 05 16:33:33     batch_size=BATCH_SIZE, use_gpu=False)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 212, in run_model_test
May 05 16:33:33     operator_export_type=operator_export_type)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 196, in run_actual_test
May 05 16:33:33     operator_export_type=operator_export_type)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/verify.py", line 356, in verify
May 05 16:33:33     operator_export_type=operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 28, in _export
May 05 16:33:33     result = utils._export(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 682, in _export
May 05 16:33:33     dynamic_axes=dynamic_axes)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 454, in _model_to_graph
May 05 16:33:33     module=module)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 200, in _optimize_graph
May 05 16:33:33     graph = torch._C._jit_pass_onnx(graph, operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
May 05 16:33:33     return utils._run_symbolic_function(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 981, in _run_symbolic_function
May 05 16:33:33     return symbolic_fn(g, *inputs, **attrs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/symbolic_helper.py", line 134, in wrapper
May 05 16:33:33     assert len(arg_descriptors) >= len(args)
May 05 16:33:33 AssertionError
May 05 16:33:33 ____________________ TestCaffe2Backend_opset10.test_scatter ____________________
May 05 16:33:33 Traceback (most recent call last):
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_common.py", line 49, in wrapper
May 05 16:33:33     return func(self)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 1616, in test_scatter
May 05 16:33:33     batch_size=BATCH_SIZE, use_gpu=False)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 216, in run_model_test
May 05 16:33:33     operator_export_type=operator_export_type)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 161, in run_debug_test
May 05 16:33:33     operator_export_type=operator_export_type)
May 05 16:33:33   File "/var/lib/jenkins/workspace/test/onnx/test_pytorch_onnx_caffe2.py", line 84, in do_export
May 05 16:33:33     out = torch.onnx._export(model, inputs, f, *args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 28, in _export
May 05 16:33:33     result = utils._export(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 682, in _export
May 05 16:33:33     dynamic_axes=dynamic_axes)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 454, in _model_to_graph
May 05 16:33:33     module=module)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 200, in _optimize_graph
May 05 16:33:33     graph = torch._C._jit_pass_onnx(graph, operator_export_type)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
May 05 16:33:33     return utils._run_symbolic_function(*args, **kwargs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/utils.py", line 981, in _run_symbolic_function
May 05 16:33:33     return symbolic_fn(g, *inputs, **attrs)
May 05 16:33:33   File "/opt/conda/lib/python3.6/site-packages/torch/onnx/symbolic_helper.py", line 134, in wrapper
May 05 16:33:33     assert len(arg_descriptors) >= len(args)
May 05 16:33:33 AssertionError
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3250,7 +3248,17 @@ def _gather(shape, index_dim, max_indices):
-    return [SampleInput(tensor, args=args) for tensor, args in test_cases]
+    samples = []
+    for tensor, args in test_cases:
+        samples.append(SampleInput(tensor, args=args, kwargs={'reduce': None}))
```
Neat; would you `.clone().detach()` `tensor` for the two extra sample inputs so each sample input has an independent `input` tensor?

mruberry reviewed (commented):
Testing updates look really good; I made one small comment. cc @anjali411 for the complex autograd extension

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1019,13 +1019,13 @@
+  self: scatter_self_backward(grad.clone(), dim, index, src, reduce)
   index: non_differentiable
-  src: grad.gather(dim, index)
+  src: scatter_src_backward(grad.clone(), self, dim, index, reduce)
```
Note that this formula has an important side effect.
When it is used for the inplace `scatter_()` (which is the one that everyone use), then we have to save the original value of `self` for this formula. This means that this formula implies a full clone of `self` every time. It would be best if we can replace this `self` by `result` as it would avoid a clone.

ysiraichi replied:
That's doesn't look very good, indeed.
However, I'm not sure how to do it without `self`. The reason is that when `reduce='multiply'` and one of the values of `src` is zero, we can't really retrieve what the value of `self` was at that position, since `result` at that position will also be zero. Any suggestions?

albanD replied:
I guess you may solve this by having two overloads: one that doesn't have the reduce argument and one that has a mandatory reduce argument.
Then you can specify formulas for each independently and save different things.

ysiraichi replied:
I guess you are right. That would also "solve" the issue with ONNX...

anjali411 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +68,67 @@
+
+  if (reduce.has_value()) {
+    TORCH_CHECK(
+        at::isFloatingType(self.scalar_type()) ||
```
why is this outrightly disallowed for integers?

ysiraichi replied:
I believe it's a problem on the CUDA side. Apparently, it depends on `gpuAtomicMul` implemented only for floating point types (not sure why, though) in [_THCAtomics.cuh_][1]. That said, I think I could, at least, enable it for CPUs.

[1]: https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCAtomics.cuh

ysiraichi committed:
Clone input tensor for different samples.

ysiraichi committed:
Revert derivatives introduction.

ysiraichi commented:
@mruberry @albanD @anjali411
I have made some changes with your feedback. Here is a brief summary:
- Cloning (and detatching) tensors for different input samples
- Revert the derivatives I had added for the `reduce` parameter (leaving them for a future PR)
- Split the `scatter` reduce variants
- Fix scalar type checks for:
    - allowing all types, including `bfloat16` for CPUs. It wasn't listed as a dispatch option, and I saw no reason why it shouldn't (I'm not really confident about this change, so please let me know if I missed something here)
    - rejecting `complex` types when `reduce='multiply'` for CUDA (same reason for integers)
- Remove a few `dimname` variants (they are back to what they were before I added more)

https://github.com/pytorch/pytorch/blob/e71b526e7ef76ce4095e4e1f5622c0c87faeb01c/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L150-L152

ysiraichi committed:
Add better type-checks.

ysiraichi committed:
Refactor type checks.

- use only `TORCH_CHECK` because of test_ops.py
- move `reduce` parameter to the end of the C++ calls
- fix test_torch.py tests

ysiraichi committed:
Make 'scatter_add' an structured kernel.

ysiraichi commented:
In these last commits, I have:
- made `scatter_add` into an structured kernel, since `scatter_add` needed the same checks as `scatter`
- removed duplicated checks in the dispatched kernels (those files in _native/{cpu,cuda}_)

ysiraichi committed:
Fix wrong behavior in meta function.

- Revert test for no-op behavior.

ysiraichi commented:
@mruberry @albanD @anjali411
I think this PR is ready for another round of reviews. Do you have some time to review it?

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,97 @@
+        op == native::SCATTER_GATHER_OP::REDUCE_MULTIPLY) {
+      TORCH_CHECK(
+          at::isFloatingType(self.scalar_type()),
+          "scatter(): Expected floating type for self when reduce=multiply.");
```
Nice error message formatting

mruberry commented on test/test_torch.py:
```diff
@@ -5490,12 +5493,15 @@ def test_put_empty(self, device):
                     self.assertEqual(dst, dst.put_(indices, src, accumulate=accumulate))
 
+    def scatter_allow_reduce(self, device, dtype, reduceop):
+        return device == 'cpu' or (device == 'cuda' and reduceop == 'multiply' and not dtype.is_floating_point)
```
The device type needs to be acquired first here:

```
device_type = torch.device(device).type
return device_type == 'cpu' ...
```

This is because the test suite supports running on specific CUDA devices, so the `device` argument to a test might be "cuda:0" or "cuda:1", not "cuda".

mruberry reviewed (commented):
Cool! Removing my request for changes. I made a few inline comments, but my concerns about testing have been addressed. Thanks @ysiraichi.

Now you just have to deal with @albanD and @ezyang. @ezyang you're coordinating with @JackCaoG, right?

ysiraichi committed:
Get device type instead of raw device.

ezyang commented:
xla companion already exists at https://github.com/pytorch/xla/pull/2936

ezyang commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -113,10 +113,7 @@ struct cpu_scatter_gather_base_kernel {
-    }
-    else {
+    if (!is_scatter_like) {
       gather_shape_check(self, dim, index, self);
```
I wonder why the gather shape checks couldn't get eliminated by structured kernels

ysiraichi replied:
I guess this one specifically can be eliminated. The other one, however, cannot (line 222). That's because `gather` (not structured) is still using that code.

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1010,85 +1101,91 @@ Tensor gather_backward(const Tensor& grad, const Tensor& self, int64_t dim, cons
+ const Tensor& out) {
+  auto mut_out = const_cast<Tensor&>(out);
+
+  if (&self != &mut_out) {
```
nyet, don't do it this way. If I do `Tensor mut_out = self;` this comparison will fail but they are still the same tensor. Instead, test `!self.is_same(mut_out)`

ysiraichi replied:
Totally agree. Didn't feel good about myself when I added that condition there either. Thanks for the suggestion (didn't know about `is_same`).

ezyang commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -215,10 +212,7 @@ struct cpu_scatter_gather_base_kernel {
     dim = maybe_wrap_dim(dim, self.dim());
 
     scatter_gather_dtype_check(method_name, self, index, src);
```
It feels like you should be able to dispense with the dtype checks now too, as they're done in the meta function

ysiraichi replied:
The problem is that (as mentioned above) `gather` uses it. If we make `gather` structured, it can go away, though.

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,97 @@
+  }
+
+  at::assert_no_internal_overlap(self);
+  at::assert_no_overlap(self, index);
```
The overlap tests here look a little suspicious. Remember that you are writing the logic which will handle BOTH functional and inplace scatter variants. So if you have logic that doesn't test `maybe_get_output()`, then it needs to be universal in both cases.  Why is it necessary for self to not have internal overlap? Internal overlap is bad if we're scattering into the tensor: weird shit will happen. But if we're scattering into self *out of place*, overlap isn't a problem, since we're going to discard the overlap when we do the initial copy. We can see this in the old implementation too:

```
Tensor scatter(const Tensor &amp; self, int64_t dim, const Tensor &amp; index, const Scalar&amp; source) {  return self.clone(at::MemoryFormat::Preserve).scatter_(dim, index, source);
```

although `scatter_` is going to do the internal overlap test, it's always going to pass, because `clone()` has the invariant that it always gets rid of internal overlap.

ysiraichi replied:
That makes sense. I hadn't really considered that. Thanks for pointing out!

ysiraichi committed:
Fix overlap checks and reduce test conditions.

ysiraichi committed:
Replace pointer comparison with `is_same`.

ysiraichi commented:
Most (11 of them) of the CI failures are tagged as "Infrastructure Fail". The other 2 don't seem related to this PR.

ysiraichi committed:
Fix return in void impl function.

ysiraichi commented:
Apart from the XLA errors, those other two doesn't seem related to this PR.
@ezyang @albanD I think this PR is ready for another round of reviews. Do you have some time to review it?

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,99 @@
+namespace at {
+namespace meta {
+
+native::SCATTER_GATHER_OP get_operator_enum(const c10::string_view reduce) {
```
Really good use case for #55526 since string comparison is nontrivially expensive

ezyang replied:
(no action needed from you)

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,99 @@
+  meta.set_output(self.sizes(), self.options());
+
+  auto output = meta.maybe_get_output(0);
+  at::assert_no_internal_overlap(output);
```
Yeah OK, this is acceptable. It is a little wasteful when the output is freshly allocated, so a very modest improvement you could do is swing this line before the `set_output` call and test if `output` is defined, and only do the overlap tests if it is defined (if it is not defined, it means we are going to freshly allocate and you don't need to test for overlap as it is guaranteed not to occur), but I'm not going to block the PR on this if there are no other substantive problems.

ysiraichi replied:
You are right. I didn't really stop to think about the complexity of these checks.
Since I have no strong reasons not to do so, I will implement that change!

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,99 @@
+  at::assert_no_overlap(output, index);
+
+  if (src.has_value()) {
+    at::assert_no_overlap(output, src.value());
```
This is a preexisting condition (no action needed), but huh, didn't realize overlap wasn't OK on src lol

bdhirsh commented on tools/autograd/gen_variable_type.py:
```diff
@@ -100,7 +100,7 @@
     'diag', 'masked_scatter', 'masked_select', 'index_fill', 'trace', 'polar', 'cumsum', 'rsub',
     'eig', 'lerp', 'linalg_vector_norm', 'cumprod', 'prod', 'index_copy', 'lu', 'unfold', 'unfold_backward',
-    'index', 'masked_fill', 'cross', 'lu_unpack'
+    'index', 'masked_fill', 'cross', 'lu_unpack', 'scatter', 'scatter_add'
```
@ysiraichi If you're wondering why you needed to add this, it looks like before this PR, `at::add_scatter` was implemented as a composite op that delegated work to `at::scatter_add_`:
```
Tensor scatter_add(const Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
    return self.clone(at::MemoryFormat::Preserve).scatter_add_(dim, index, source);
 }
```

And that includes delegating autograd. It looks like `scatter_add_` supports complex autograd, which means that `scatter_add` inherited that support too.

Now that `scatter_add` is structured, it needs its own autograd support. It looks like you added that in `derivatives.yaml`, but it also looks like the [codegen is set up](https://github.com/pytorch/pytorch/blob/12bb1e86ed150cfed4b6a7496db982ab6ce3450e/tools/autograd/gen_variable_type.py#L561) so that operators with specific return types (including `Tensor`) don't automatically get complex autograd support; you have to add them to this list once you're convinced that the implementation works with complex autograd.

ysiraichi replied:
Ah, you are totally right!
I hadn't seen that `differentiable_OUTPUTS`... That makes total sense now!
Thanks for that.

ysiraichi committed:
Move overlap checks after checking if output is defined.

ysiraichi commented:
The 2 CI failures aren't related to this PR.

ezyang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -68,7 +69,99 @@
+      TORCH_CHECK(
+          self.scalar_type() != kBFloat16,
+          "scatter(): bfloat16 not supported on cpu")
+    }
```
This works, but it might be better to put the CPU/CUDA specific checks in their own kernels (because you can think of it as a backend specific implementation detail, as opposed to something that is backend generic)

ysiraichi replied:
That does make sense. Since the kernels are using `AT_DISPATCH_ALL_TYPES...` (which will error out on not supported types), I guess I could simply remove these checks in the meta function. What do you think?

ezyang replied:
Yeah, that's what I would do in that case

ezyang reviewed (approved):
Thanks, nice work

ezyang commented:
Hmm, needs XLA update. cc @JackCaoG

ezyang commented:
```
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatter
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 [  FAILED  ] AtenXlaTensorTest.TestScatter (32 ms)
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatterR1
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 [  FAILED  ] AtenXlaTensorTest.TestScatterR1 (35 ms)
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatterR3
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 [  FAILED  ] AtenXlaTensorTest.TestScatterR3 (30 ms)
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatterBiggerSource
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 [  FAILED  ] AtenXlaTensorTest.TestScatterBiggerSource (28 ms)
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatterScalar
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 [  FAILED  ] AtenXlaTensorTest.TestScatterScalar (18 ms)
May 26 13:16:42 [ RUN      ] AtenXlaTensorTest.TestScatterAdd
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:50: Failure
May 26 13:16:42 Value of: changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
May 26 13:16:42 /var/lib/jenkins/workspace/xla/test/cpp/torch_xla_test.cpp:59: Failure
May 26 13:16:42 Value of: !changed.empty()
May 26 13:16:42   Actual: false
May 26 13:16:42 Expected: true
```

JackCaoG commented:
@ezyang xla pr is in https://github.com/pytorch/xla/pull/2936, let me rerun the CI on that pr (and maybe rebase) to make sure test still passes.

ezyang commented:
oops, me and my bird brain. Thanks for reminding me :)

JackCaoG commented:
I need a bit more time to rebase the pt/xla pr since it is a bit old. I will report back when that is ready.

JackCaoG commented:
@ysiraichi Can you rebase this pr so pt/xla can pr can build? I manually patch your pr to my local build and everything seems fine but I still want the CI of pt/xla pr to pass.

ysiraichi committed:
Remove device specific checks.

ysiraichi committed:
Fix tests not supporting `bfloat16`.

ysiraichi committed:
Run test only on CUDA (there are no unsupported types on CPU anymore)

ysiraichi committed:
Fix `@onlyOnCUDA` to `@onlyCUDA`.

JackCaoG commented:
@ezyang pt/xla pr is ready. If you give me a ping before merging this pr I can remove the pin and merge the pt/xla pr.

ezyang committed:
Merge remote-tracking branch 'origin/master' into scatter-reduce-kwarg
